huaweicloud/services/apig/data_source_huaweicloud_apig_environments.go

```diff
@@ -0,0 +1,116 @@
+package apig
+
+import (
+	"context"
+	"regexp"
+
+	"github.com/chnsz/golangsdk/openstack/apigw/dedicated/v2/environments"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/helper/hashcode"
+)
+
+func DataSourceEnvironments() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceEnvironmentsRead,
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"instance_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ValidateFunc: validation.All(
+					validation.StringMatch(regexp.MustCompile(`^[A-Za-z]([\w0-9]+)?$`),
+						"The name can only consist of letters, digits and underscores (_), and must start with a letter."),
+					validation.StringLenBetween(3, 64),
+				),
+			},
+			"environments": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"description": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"create_time": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func buildListOpts(d *schema.ResourceData) environments.ListOpts {
+	return environments.ListOpts{
+		Name: d.Get("name").(string),
+	}
+}
```

Comment:
this func looks not very necessary because there is only one parameter and the struct of ListOpts is quite simple

Reply:
done, this function has been removed.